main.go

```diff
@@ -0,0 +1,517 @@
+		} else {
+			f.WriteString("Rows that generated conversion errors:\n")
+		}
+		for _, r := range l {
+			_, err := f.WriteString("  " + r + "\n")
+			if err != nil {
+				fmt.Printf("Can't write out bad data file: %v\n", err)
+				return
+			}
+		}
+	}
+	if badWrites > 0 {
+		l := bw.SampleBadRows(maxRows)
+		if int64(len(l)) < badWrites {
+			f.WriteString("A sample of rows that successfully converted but couldn't be written to Spanner:\n")
+		} else {
+			f.WriteString("Rows that successfully converted but couldn't be written to Spanner:\n")
+		}
+		for _, r := range l {
+			_, err := f.WriteString("  " + r + "\n")
+			if err != nil {
+				fmt.Printf("Can't write out bad data file: %v\n", err)
+				return
+			}
+		}
+	}
+	fmt.Printf("See file '%s' for details of bad rows\n", name)
+}
+
+func getDatabaseName(now time.Time) (string, error) {
+	if dbNameOverride != "" {
+		return dbNameOverride, nil
+	}
+	return generateName(fmt.Sprintf("pg_dump_%s", now.Format("2006-01-02")))
+}
+
+// analyzeError inspects an error returned from Cloud Spanner and adds information
+// about potential root causes e.g. authentication issues.
+func analyzeError(err error, project string) error {
+	e := strings.ToLower(err.Error())
+	if containsAny(e, []string{"unauthenticated", "cannot fetch token", "default credentials"}) {
+		return fmt.Errorf("%w.\n"+`
+Possible cause: credentials are mis-configured. Do you need to run
+
+  gcloud auth application-default login
+
+or configure environment variable GOOGLE_APPLICATION_CREDENTIALS.
+See https://cloud.google.com/docs/authentication/getting-started.
+`, err)
+	}
+	if containsAny(e, []string{"instance not found"}) && instanceOverride != "" {
+		return fmt.Errorf("%w.\n"+`
+Possible cause: Spanner instance specified via instance option does not exist.
+Please check that '%s' is correct and that it is a valid Spanner
+instance for project %s.
+`, err, instanceOverride, project)
+	}
+	return err
+}
+
+func usage() {
+	fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+	flag.PrintDefaults()
+	fmt.Fprintf(os.Stderr, `Note: input is always read from stdin.
+Sample usage:
+  pg_dump mydb | %s
+  %s < my_pg_dump_file
+`, os.Args[0], os.Args[0])
+}
+
+func containsAny(s string, l []string) bool {
+	for _, a := range l {
+		if strings.Contains(s, a) {
+			return true
+		}
+	}
+	return false
+}
+
+func generateName(prefix string) (string, error) {
+	b := make([]byte, 4)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", fmt.Errorf("error generating name: %w", err)
+
+	}
+	return fmt.Sprintf("%s_%x-%x", prefix, b[0:2], b[2:4]), nil
+}
+
+func getClient(db string) (*sp.Client, error) {
+	ctx := context.Background()
+	return sp.NewClient(ctx, db)
+}
+
+func getSize(f *os.File) (int64, error) {
+	info, err := f.Stat()
+	if err != nil {
+		return 0, fmt.Errorf("can't stat file: %w", err)
+	}
+	return info.Size(), nil
+}
+
+func setFilePrefix(dbName string) {
```

Comment:
I'm not sure this makes sense to be a separate function. As it is, I think it could be confusing as by convention, the function implies that you're setting the filePrefix to dbName.

I would either remove this function or reduce it to:
`func setFilePrefix(prefix string) {
        filePrefix = prefix
}`
and change the call to:
`prefix := dbName + "."`
`if filePrefix != "" {
        prefix = filePrefix
}`
`setFilePrefix(prefix)`
That way it's clear what the default is and that the flag is overriding it

Reply:
Function removed (inlined).